### ISSUE TITLE
Can you help me figure out how to add serde to this struct?

### ISSUE THREAD

**Issue by pitdicker** [serde-rs/serde#1207]
In https://github.com/rust-lang-nursery/rand/pull/325 I am trying to add serde to struct, with no luck. I am kind of lost in the trait bound errors. Can you help me out?
By now I have serialization working, but no idea where to start with deserialization.

Is is about a struct where one field is a generic:
```rust
pub struct BlockRng<R: BlockRngCore + ?Sized> {
    pub results: R::Results,
    pub index: usize,
    pub core: R,
}

pub trait BlockRngCore {
    type Item;

    /// Results type. This is the 'block' an RNG implementing `BlockRngCore`
    /// generates, which will usually be an array like `[u32; 16]`.
    type Results: AsRef<[Self::Item]> + AsMut<[Self::Item]> + Default;

    /// Generate a new block of results.
    fn generate(&mut self, results: &mut Self::Results);
}
```
Please see the linked PR for my latest attempt.

**Comment by dtolnay** [comment]
You may need to use a [handwritten generic type bound](https://serde.rs/attr-bound.html) to enforce that R::Results is serializable and deserializable.

```rust
#[derive(Serialize, Deserialize)]
pub struct BlockRng<R: BlockRngCore + ?Sized> {
    #[serde(bound(
        serialize = "R::Results: Serialize",
        deserialize = "R::Results: Deserialize<'de>"))]
    pub results: R::Results,
    pub index: usize,
    pub core: R,
}
```

**Comment by pitdicker** [comment]
Thank you! I'll try that soon, but it would have taken me hours more to find, even with the good documentation!

**Comment by pitdicker** [comment]
It works 😄 .

**Comment by dtolnay** [comment]
This has been fixed in Serde 1.0.38 so the original code should work as written and the bounds on the associated type can be inferred.

```rust
#[derive(Serialize, Deserialize)]
struct BlockRng<R: BlockRngCore + ?Sized> {
    results: R::Results,
    index: usize,
    core: R,
}
```

**Comment by pitdicker** [comment]
Wow, that is nice!
